src/Raven.Client/Document/Subscription.cs

```diff
@@ -32,289 +40,440 @@ namespace Raven.Client.Document
 
+
+        /// <summary>
+        ///     It indicates if the subscription is in errored state because one of subscribers threw an exception.
+        /// </summary>
+        public bool IsErroredBecauseOfSubscriber { get; private set; }
+
+        /// <summary>
+        ///     The last exception thrown by one of subscribers.
+        /// </summary>
+        public Exception LastSubscriberException { get; private set; }
+
+        /// <summary>
+        ///     The last subscription connection exception.
+        /// </summary>
+        public Exception SubscriptionConnectionException { get; private set; }
+
+        /// <summary>
+        ///     It determines if the subscription connection is closed.
+        /// </summary>
+        public bool IsConnectionClosed { get; private set; }
+
+        public void Dispose()
+        {
+            if (disposed)
+                return;
+
+            DisposeAsync().Wait();
+        }
+
+        public Task DisposeAsync()
+        {
+            if (disposed)
+                return new CompletedTask();
+
+            disposed = true;
+
+            OnCompletedNotification();
+
+            subscribers.Clear();
+
+            if (endedBulkInsertsObserver != null)
+                endedBulkInsertsObserver.Dispose();
+
+            if (dataSubscriptionReleasedObserver != null)
+                dataSubscriptionReleasedObserver.Dispose();
+
+            cts.Cancel();
+
+            anySubscriber.Set();
+
+            foreach (var task in new[] { pullingTask, startPullingTask })
+            {
+                if (task == null)
+                    continue;
+
+                switch (task.Status)
+                {
+                    case TaskStatus.RanToCompletion:
+                    case TaskStatus.Canceled:
+                        break;
+                    default:
+                        try
+                        {
+                            task.Wait();
+                        }
+                        catch (AggregateException ae)
+                        {
+                            if (ae.InnerException is OperationCanceledException == false &&
+                                ae.InnerException is WebSocketException == false)
+                            {
+                                throw;
+                            }
+                        }
+
+                        break;
+                }
+            }
+
+            if (IsConnectionClosed)
+                return new CompletedTask();
+
+            return CloseSubscription();
+        }
+
+        public IDisposable Subscribe(IObserver<T> observer)
+        {
+            if (IsErroredBecauseOfSubscriber)
+                throw new InvalidOperationException(
+                    "Subscription encountered errors and stopped. Cannot add any subscriber.");
+
+            if (subscribers.TryAdd(observer))
+            {
+                if (subscribers.Count == 1)
+                    anySubscriber.Set();
```

Comment:
Bad, race condition here. If you have two users that can register on the same time, you might get both out at the same time